drivers/shared/executor/executor_linux_test.go

```diff
@@ -306,41 +306,92 @@ func TestUniversalExecutor_LookupTaskBin(t *testing.T) {
 	t.Parallel()
 	require := require.New(t)
 
-	// Create a temp dir
-	tmpDir, err := ioutil.TempDir("", "")
-	require.Nil(err)
-	defer os.Remove(tmpDir)
-
-	// Create the command
-	cmd := &ExecCommand{Env: []string{"PATH=/bin"}, TaskDir: tmpDir}
-
-	// Make a foo subdir
-	os.MkdirAll(filepath.Join(tmpDir, "foo"), 0700)
+	cases := []struct {
+		name             string
+		cmd              string
+		env              string
+		shouldErr        bool
+		writeTaskDirPath string
+	}{
+		{
+			name:             "reference to a binary that exists in taskdir root",
+			cmd:              "some-root-bin",
+			writeTaskDirPath: "/some-root-bin",
+		},
+		{
+			name:      "reference to a binary that doesn't exist in taskdir root",
+			cmd:       "some-root-bin",
+			shouldErr: true,
+		},
+		{
+			name:             "absolute reference to a binary that exists in taskdir",
+			cmd:              "/abs/some-bin",
+			writeTaskDirPath: "/abs/some-bin",
+		},
+		{
+			name:      "absolute reference to a binary that doesn't exist in taskdir",
+			cmd:       "/abs/some-bin",
+			shouldErr: true,
+		},
+		{
+			name:             "relative reference to a binary that exists in taskdir/bin",
+			cmd:              "some-bin",
+			writeTaskDirPath: "/bin/some-bin",
+		},
+		{
+			name:             "relative reference to a binary that exists in taskdir/usr/bin",
+			cmd:              "some-bin",
+			writeTaskDirPath: "/usr/bin/some-bin",
+		},
+		{
+			name:             "relative reference to a binary that exists in taskdir/usr/local/bin",
+			cmd:              "some-bin",
+			writeTaskDirPath: "/usr/local/bin/some-bin",
+		},
+		{
+			name:             "relative reference to a binary that exists in PATH-defined dir",
+			cmd:              "some-bin",
+			env:              "PATH=/usr/local/bin:/bar",
+			writeTaskDirPath: "/bar/some-bin",
+		},
+		{
+			name:      "relative reference to a binary that doesn't exist",
+			cmd:       "some-bin",
+			shouldErr: true,
+		},
```

Comment:
The [caller](https://github.com/hashicorp/nomad/blob/v1.0.1/drivers/shared/executor/executor_linux.go#L120)  for `lookupTaskBin` bails out if we return an error. Which implies that if we pass the task an absolute path to a binary that exists but is not on PATH, it will fail. I just verified that's the case:

> `2021-01-12T20:18:25.990Z [ERROR] client.alloc_runner.task_runner: running driver failed: alloc_id=c0b5da7c-c0ba-f10a-7156-69e0dcab55a7 task=exectask error="failed to launch command with executor: rpc error: code = Unknown desc = file /opt/whatever.sh not found under path /tmp/NomadClient407725986/c0b5da7c-c0ba-f10a-7156-69e0dcab55a7/exectask"`

That doesn't _seem_ correct though, and certainly isn't documented. Maybe we want to fix it or at least document the behavior? (Or maybe we can just open a new issue for the problem to make sure we come back to it later.)

Either way, I think we should have a test exercising the expectation here.

Reply:
The case you describe should be covered by the test on 327, an absolute reference to a binary that exists in taskdir. PATH shouldn't come into play there. Can you provide me your jobspec?

Reply:
From the current master, assuming there's an executable script at `/opt/whatever.sh`:

```hcl
job "example" {
  datacenters = ["dc1"]

  group "group" {

    task "task" {
      driver = "exec"

      config {
        command = "/opt/whatever.sh"
      }
    }
  }
}
```

Results in:

```
$ nomad alloc status 846
...
Recent Events:
Time                  Type             Description
2021-01-12T20:55:07Z  Killing          Sent interrupt. Waiting 5s before force killing
2021-01-12T20:55:05Z  Alloc Unhealthy  Unhealthy because of failed task
2021-01-12T20:55:05Z  Not Restarting   Error was unrecoverable
2021-01-12T20:55:05Z  Driver Failure   failed to launch command with executor: rpc error: code = Unknown desc = file /opt/whatever.sh not found under path /tmp/NomadClient392677841/84637dd7-245a-04cf-f452-2b9e7ffac75d/task
2021-01-12T20:55:02Z  Task Setup       Building Task Directory
2021-01-12T20:55:02Z  Received         Task received by client

$ ls -lah /opt/whatever.sh
-rwxr-xr-x 1 root root 39 Jan 12 20:18 /opt/whatever.sh
```

Reply:
Your job does work if the driver is `raw_exec`, however.

Per the `exec` docs on `command`:

> The command to execute. Must be provided. If executing a binary that exists on the host, **the path must be absolute and within the task's chroot.**

So the error you're seeing is expected for an `exec` task.

I think the simplest thing here is to just remove the `PATH=` code, since it's undocumented (as far as I can tell) and didn't work anyway.

Reply:
Ah, I missed the `raw_exec`/`exec` behavior difference. Well that means this is fixed then, right? Why not keep it?

Reply:
It's fixed, but it's also undocumented. It's easier to just delete it. I made it two separate commits with the idea that either one could get merged; after looking at the actual use case for this I think it's not very compelling, so I'd rather not add support now.

Reply:
LGTM